File: euler_to_transform.m
function T = euler_to_transform(x,y,z,yaw,pitch,roll)
%EULER_TO_TRANSFORM: converts yawpitchr euler angles and a translation vector to
%create a trasnformation matrix between the two

%Input is how the frame we want to go FROM looks in the frame we want to GO
%TO
%%{B} -> {A} refer to "Introduction to Robotics: Mechanics and Control" fig
%2.7 for drawn example

%%Create the rotation matrix found in eq 2.71
rotation = [cos(yaw)*cos(pitch), cos(yaw)*sin(pitch)*sin(roll)-sin(yaw)*cos(roll),cos(yaw)*sin(pitch)*cos(roll)+sin(yaw)*sin(roll);
            sin(yaw)*cos(pitch), sin(yaw)*sin(pitch)*sin(roll)+cos(yaw)*cos(roll), sin(yaw)*sin(pitch)*cos(roll)-cos(yaw)*sin(roll);
            -sin(pitch),cos(pitch)*sin(roll),cos(pitch)*cos(roll)];
trans = [x;y;z]; %%Create translation vector

T=[rotation,trans;zeros(1,3),1]; %%Attach matrices together using format found in eq 2.19
end

File: Euler_to_Transform.py
if __name__ == "__main__": #Unit test data
    import math
    x = 10
    y = 5
    d = 0
    yaw = 30*math.pi/180
    pitch = 0
    roll = 0

def Euler_to_Transform(x,y,z,yaw,pitch,roll):
    #Function to calculate the transformation between two coordinate
    import math #Grab math library to calculate cos(x) and sin(x)
    import numpy as np

    #Pre-calculate the required numbers to make code shorter
    cy=math.cos(yaw)
    cp=math.cos(pitch)
    cr=math.cos(roll)
    sy=math.sin(yaw)
    sp=math.sin(pitch)
    sr=math.sin(roll)

    #Calculate a create the transform 
    T=np.array([[cy*cp, cy*sp*sr-sy*cr,cy*sp*cr+sy*sr,x],
               [sy*cp,sy*sp*sr+cy*cr,sy*sp*cr-cy*sr,y],
               [-1*sp,cp*sr,cp*cr,z],
               [0,0,0,1]],dtype=float)

    return T #return the transform

